From 4f5c9f33ade060fd2407742ba40abd6f720cefa7 Mon Sep 17 00:00:00 2001
From: Yanjiang Jin <yanjiang.jin@windriver.com>
Date: Fri, 28 Oct 2011 15:06:46 +0800
Subject: [PATCH] mips/nlm_xlr_atx_64_be: fix the definition of cache coherency bit

SDK use the fixed offset value in the three cache attributes macro.
It's incorrect when we disable the HUGETLB_PAGE.

The _CACHE_CACHABLE_NONCOHERENT affects the PAGE_NONE macro in
protection_map[] which is the mapping table of vm and hardware
protection attributes. The kernel would get incorrect attribute
bits when using the table.

The cache attribute becomes 0x7 on mprotect() system call, but 0x7 is
illegal in XLS/XLR platform and it will cause kernel unaligned access.

Unhandled kernel unaligned access[#1]:
Cpu 28
$ 0   : 0000000000000000 0000000000000001 0000000000000000 ffffff0000000000
......
$28   : a80000012201c000 a80000012201fbe0 0000000000420000 ffffffff8342049c
Hi    : 000000000000026b
Lo    : 00000000000001b6
epc   : ffffffff834202b8 emulate_load_store_insn+0x3b0/0x470
    Not tainted
ra    : ffffffff8342049c do_ade+0x124/0x2b0
Status: 1000dce3    KX SX UX KERNEL EXL IE
Cause : 00800010
BadVA : 000000002aac8000
PrId  : 000c0009 (XLR732 Rev C4)
Modules linked in:
Process vsftpd (pid: 1429, threadinfo=a80000012201c000,
	task=a80000012572b838, tls=000000002b0ed570)
Stack : a800000083f2f400 ffffffff8350ab60 0000000000000001 ffffffff8354c2d4
        a800000123feeb40 ffffffff8340fa3c 00000000906c0397 ffffffff83522d5c
        ...
Call Trace:
[<ffffffff834202b8>] emulate_load_store_insn+0x3b0/0x470
[<ffffffff8342049c>] do_ade+0x124/0x2b0
[<ffffffff83400484>] ret_from_exception+0x0/0x10

Code: 00431024  5440ff4c  de030100 <8a230000> 9a230003  24020000
	1440ff60  0060882d  08d0805e
Disabling lock debugging due to kernel taint

Segmentation fault

Signed-off-by: Yanjiang Jin<yanjiang.jin@windriver.com>
---
 arch/mips/include/asm/pgtable-bits.h |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/arch/mips/include/asm/pgtable-bits.h b/arch/mips/include/asm/pgtable-bits.h
index cccecd0..fdd3784 100644
--- a/arch/mips/include/asm/pgtable-bits.h
+++ b/arch/mips/include/asm/pgtable-bits.h
@@ -194,9 +194,9 @@ static inline uint64_t pte_to_entrylo(unsigned long pte_val)
 
 #elif defined(CONFIG_CPU_PHOENIX)
 
-#define _CACHE_UNCACHED             (2<<9)
-#define _CACHE_CACHABLE_COW         (3<<9)
-#define _CACHE_CACHABLE_NONCOHERENT (3<<9)
+#define _CACHE_UNCACHED             (2<<_CACHE_SHIFT)
+#define _CACHE_CACHABLE_COW         (3<<_CACHE_SHIFT)
+#define _CACHE_CACHABLE_NONCOHERENT (3<<_CACHE_SHIFT)
 
 #elif defined(CONFIG_CPU_RM9000)
 
-- 
1.7.0.4

